Frameworks/include/CGImageInternal.h

```diff
@@ -96,31 +105,267 @@ typedef enum {
+    /*Indexed*/
+    { GUID_WICPixelFormat1bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 1 } },
+    { GUID_WICPixelFormat2bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 2 } },
+    { GUID_WICPixelFormat4bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 4 } },
+    { GUID_WICPixelFormat8bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 8 } }
+};
+
+struct __CGImageImpl {
+    Microsoft::WRL::ComPtr<IWICBitmapSource> bitmapImageSource;
+    bool isMask;
+    bool interpolate;
+    woc::unique_cf<CGColorSpaceRef> colorSpace;
+    CGImageAlphaInfo alphaInfo;
+    size_t height;
+    size_t width;
+    size_t bitsPerPixel;
+    size_t bitsPerComponent;
+    CGBitmapInfo bitmapInfo;
+    CGColorRenderingIntent renderingIntent;
+
+    __CGImageImpl() {
+        height = 0;
+        width = 0;
+        bitsPerComponent = 0;
+        bitsPerPixel = 0;
+        bitmapInfo = kCGBitmapByteOrderDefault;
+        alphaInfo = kCGImageAlphaNone;
+        isMask = false;
+        interpolate = false;
+        renderingIntent = kCGRenderingIntentDefault;
+    }
+
+    inline WICPixelFormatGUID PixelFormat() const {
+        WICPixelFormatGUID pixelFormat;
+        RETURN_RESULT_IF_FAILED(bitmapImageSource->GetPixelFormat(&pixelFormat), GUID_WICPixelFormatUndefined);
+        return pixelFormat;
+    }
+
+    inline const __CGImagePixelProperties* Properties() const {
+        WICPixelFormatGUID pixelFormat = PixelFormat();
+        RETURN_NULL_IF(pixelFormat == GUID_WICPixelFormatUndefined);
+
+        auto iterator = s_PixelFormats.find(pixelFormat);
+        RETURN_NULL_IF(iterator == s_PixelFormats.end());
+
+        return &iterator->second;
+    }
+
+    inline size_t BitsPerPixel() const {
+        const __CGImagePixelProperties* properties = Properties();
+        RETURN_RESULT_IF_NULL(properties, 0);
+        return properties->bitsPerPixel;
+    }
+
+    inline size_t BitsPerComponent() const {
+        const __CGImagePixelProperties* properties = Properties();
+        RETURN_RESULT_IF_NULL(properties, 0);
+        return properties->bitsPerComponent;
+    }
+
+    inline CGBitmapInfo BitmapInfo() const {
+        const __CGImagePixelProperties* properties = Properties();
+        RETURN_RESULT_IF_NULL(properties, 0);
+        return properties->bitmapInfo;
+    }
+
+    inline CGImageAlphaInfo AlphaInfo() const {
+        return static_cast<CGImageAlphaInfo>(BitmapInfo() & kCGBitmapAlphaInfoMask);
+    }
+
+    inline CGColorSpaceRef ColorSpace() {
+        const __CGImagePixelProperties* properties = Properties();
+        RETURN_NULL_IF(!properties);
+        return _CGColorSpaceCreate(properties->colorSpaceModel);
+    }
+
+    inline void SetImageSource(Microsoft::WRL::ComPtr<IWICBitmapSource> source) {
+        bitmapImageSource = source;
```

Comment:
add this to your next pull request, not this one.

---
In reply to: [84975050](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r84975050) [](ancestors = 84975050)